web/pdf_find_controller.js

```diff
@@ -331,193 +469,161 @@ class PDFFindController {
-          break;
+        if (p4) {
+          // Diacritics are removed with few exceptions.
+          return DIACRITICS_EXCEPTION.has(p4.charCodeAt(0)) ? p4 : "";
         }
-        if (
-          entireWord &&
-          !this._isEntireWord(pageContent, matchIdx, subqueryLen)
-        ) {
-          continue;
+
+        // A letter has been matched and it can be followed by any diacritics
+        // in normalized text.
+        if (hasDiacritics) {
+          isUnicode = true;
+          return `${p5}\\p{M}*`;
         }
-        const originalMatchIdx = getOriginalIndex(matchIdx, pageDiffs),
-          matchEnd = matchIdx + subqueryLen - 1,
-          originalQueryLen =
-            getOriginalIndex(matchEnd, pageDiffs) - originalMatchIdx + 1;
-
-        // Other searches do not, so we store the length.
-        matchesWithLength.push({
-          match: originalMatchIdx,
-          matchLength: originalQueryLen,
-          skipped: false,
-        });
+        return p5;
       }
-    }
+    );
 
-    // Prepare arrays for storing the matches.
-    this._pageMatchesLength[pageIndex] = [];
-    this._pageMatches[pageIndex] = [];
+    if (matchDiacritics) {
+      // aX must not match aXY.
+      if (hasDiacritics) {
+        isUnicode = true;
+        query = `${query}(?=[${DIACRITICS_EXCEPTION_STR}]|[^\\p{M}]|$)`;
+      }
+    }
 
-    // Sort `matchesWithLength`, remove intersecting terms and put the result
-    // into the two arrays.
-    this._prepareMatches(
-      matchesWithLength,
-      this._pageMatches[pageIndex],
-      this._pageMatchesLength[pageIndex]
-    );
+    return [isUnicode, query];
   }
 
   _calculateMatch(pageIndex) {
-    let pageContent = this._pageContents[pageIndex];
-    const pageDiffs = this._pageDiffs[pageIndex];
     let query = this._query;
-    const { caseSensitive, entireWord, phraseSearch } = this._state;
-
     if (query.length === 0) {
       // Do nothing: the matches should be wiped out already.
       return;
     }
 
-    if (!caseSensitive) {
-      pageContent = pageContent.toLowerCase();
-      query = query.toLowerCase();
-    }
+    const { caseSensitive, entireWord, phraseSearch } = this._state;
+    const pageContent = this._pageContents[pageIndex];
+    const hasDiacritics = this._hasDiacritics[pageIndex];
 
+    let flags = caseSensitive ? "g" : "gi";
+    let isUnicode = false;
     if (phraseSearch) {
-      this._calculatePhraseMatch(
-        query,
-        pageIndex,
-        pageContent,
-        pageDiffs,
-        entireWord
-      );
+      [isUnicode, query] = this._convertToRegExpString(query, hasDiacritics);
     } else {
-      this._calculateWordMatch(
-        query,
-        pageIndex,
-        pageContent,
-        pageDiffs,
-        entireWord
-      );
+      // Words are sorted in reverse order to be sure that "foobar" is matched
+      // before "foo" in case the query is "foobar foo".
+      query = query
+        .match(/\S+/g)
+        .sort()
```

Comment:
Would it be possible for `query` to contain only whitespaces here?
Because in that case `String.prototype.match()` will return `null` which will thus cause the `sort()`-call to throw an Error.

Reply:
Good catch, thank you.